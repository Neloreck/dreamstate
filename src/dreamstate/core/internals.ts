import type { Context } from "react";

import type {
  TAnyContextManagerConstructor,
  TAnyValue,
  TQuerySubscriptionMetadata,
  TSignalSubscriptionMetadata,
} from "@/dreamstate/types";

/**
 * Meta symbols for private internals in context managers.
 */
export const SIGNAL_METADATA_SYMBOL: unique symbol = Symbol("SIGNAL_METADATA");
export const QUERY_METADATA_SYMBOL: unique symbol = Symbol("QUERY_METADATA");
export const SIGNALING_HANDLER_SYMBOL: unique symbol = Symbol("SIGNALING_HANDLER");
export const SCOPE_SYMBOL: unique symbol = Symbol("SCOPE");

/**
 * A weak map registry that stores React context instances bound to specific manager classes.
 *
 * This registry ensures that the library does not retain unnecessary references to manager classes,
 * preventing memory leaks and unintended side effects.
 *
 * This is particularly useful in scenarios such as:
 * - Hot Module Replacement
 * - Module unloading
 * - Scope disposal
 */
export const CONTEXT_REACT_CONTEXTS_REGISTRY: WeakMap<
  TAnyContextManagerConstructor,
  Context<TAnyValue>
> = new WeakMap();

export const SIGNAL_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TSignalSubscriptionMetadata> =
  new WeakMap();

export const QUERY_METADATA_REGISTRY: WeakMap<TAnyContextManagerConstructor, TQuerySubscriptionMetadata> =
  new WeakMap();
