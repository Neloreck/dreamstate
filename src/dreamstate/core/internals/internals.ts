import type { Context } from "react";

import type {
  TServiceMap,
  TQuerySubscriptionMetadata,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TUpdateObserver,
  TUpdateSubscriber,
  TDreamstateService,
  TAnyObject
} from "@/dreamstate/types";

/**
 * Global shared constants.
 */

export const EMPTY_ARR: [] = [];

/**
 * Store related constants.
 */

// Currently active services instance types.
export const CONTEXT_SERVICES_ACTIVATED: Set<TDreamstateService> = new Set();

// Currently active services instance types.
export const CONTEXT_SERVICES_REFERENCES: TServiceMap<number> = new WeakMap();

// ReactContext registry, lazy initialized constants.
export const CONTEXT_REACT_CONTEXTS_REGISTRY: TServiceMap<Context<any>> = new WeakMap();

// ContextService instances registry.
export const CONTEXT_SERVICES_REGISTRY: TServiceMap<InstanceType<TDreamstateService>> = new WeakMap();

// ContextManagers context references registry.
export const CONTEXT_STATES_REGISTRY: TServiceMap<TAnyObject> = new WeakMap();

// ContextManagers observers registry - Providers of related context.
export const CONTEXT_OBSERVERS_REGISTRY: TServiceMap<Set<TUpdateObserver>> = new WeakMap();

// ContextManagers subscribers registry for data updates.
export const CONTEXT_SUBSCRIBERS_REGISTRY: TServiceMap<Set<TUpdateSubscriber<any>>> = new WeakMap();

// Signal handlers registry bound to context instances.
export const CONTEXT_SIGNAL_HANDLERS_REGISTRY: TServiceMap<TSignalListener<any>> = new WeakMap();

// Signal listeners metadata registry related to services.
export const CONTEXT_SIGNAL_METADATA_REGISTRY: TServiceMap<TSignalSubscriptionMetadata> = new WeakMap();

// Query listeners metadata registry related to services.
export const CONTEXT_QUERY_METADATA_REGISTRY: TServiceMap<TQuerySubscriptionMetadata> = new WeakMap();

// Global signal listeners registry.
export const SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>> = new Set();
