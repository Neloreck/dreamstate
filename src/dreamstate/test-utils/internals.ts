/**
 * Meta symbols for private internals in context managers for test-utils.
 */
export const IS_MOCKED: unique symbol = Symbol("IS_MOCKED");
