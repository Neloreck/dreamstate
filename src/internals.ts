import type { Context } from "react";

import type {
  TWorkerMap,
  TQuerySubscriptionMetadata,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TUpdateObserver,
  TUpdateSubscriber, TDreamstateWorker
} from "./types";

export const EMPTY_STRING: string = "";
export const MANAGER_REGEX: RegExp = /Manager$/;
export const EMPTY_ARR: [] = [];

export const NESTED_STORE_KEY: unique symbol = Symbol();

/**
 * Store related constants.
 */

export const CONTEXT_WORKERS_ACTIVATED: Set<TDreamstateWorker> = new Set();

export const CONTEXT_REACT_CONTEXTS_REGISTRY: TWorkerMap<Context<any>> = new WeakMap();
export const CONTEXT_WORKERS_REGISTRY: TWorkerMap<InstanceType<TDreamstateWorker>> = new WeakMap();
export const CONTEXT_STATES_REGISTRY: TWorkerMap<object> = new WeakMap();

export const CONTEXT_OBSERVERS_REGISTRY: TWorkerMap<Set<TUpdateObserver>> = new WeakMap();
export const CONTEXT_SUBSCRIBERS_REGISTRY: TWorkerMap<Set<TUpdateSubscriber<any>>> = new WeakMap();
export const CONTEXT_SIGNAL_HANDLERS_REGISTRY: TWorkerMap<TSignalListener<any>> = new WeakMap();

export const CONTEXT_SIGNAL_METADATA_REGISTRY: TWorkerMap<TSignalSubscriptionMetadata> = new WeakMap();
export const CONTEXT_QUERY_METADATA_REGISTRY: TWorkerMap<TQuerySubscriptionMetadata> = new WeakMap();

export const SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>> = new Set();

declare const IS_DEBUG: boolean;

if (IS_DEBUG) {
  // @ts-ignore debug-only declaration.
  window.__DREAMSTATE_DEBUG_INTERNALS__ = {
    CONTEXT_WORKERS_REGISTRY,
    CONTEXT_OBSERVERS_REGISTRY,
    CONTEXT_SUBSCRIBERS_REGISTRY,
    CONTEXT_STATES_REGISTRY,
    CONTEXT_SIGNAL_METADATA_REGISTRY,
    CONTEXT_SIGNAL_HANDLERS_REGISTRY,
    CONTEXT_QUERY_METADATA_REGISTRY,
    SIGNAL_LISTENERS_REGISTRY
  };
}
