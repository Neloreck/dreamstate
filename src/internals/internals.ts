import type { Context } from "react";

import type {
  TWorkerMap,
  TQuerySubscriptionMetadata,
  TSignalListener,
  TSignalSubscriptionMetadata,
  TUpdateObserver,
  TUpdateSubscriber,
  TDreamstateWorker
} from "../types";

/**
 * Global shared constants.
 */

export const EMPTY_ARR: [] = [];

/**
 * Store related constants.
 */

// Currently active workers instance types.
export const CONTEXT_WORKERS_ACTIVATED: Set<TDreamstateWorker> = new Set();

// ReactContext registry, lazy initialized constants.
export const CONTEXT_REACT_CONTEXTS_REGISTRY: TWorkerMap<Context<any>> = new WeakMap();

// ContextWorker instances registry.
export const CONTEXT_WORKERS_REGISTRY: TWorkerMap<InstanceType<TDreamstateWorker>> = new WeakMap();

// ContextManagers context references registry.
export const CONTEXT_STATES_REGISTRY: TWorkerMap<object> = new WeakMap();

// ContextManagers observers registry - Providers of related context.
export const CONTEXT_OBSERVERS_REGISTRY: TWorkerMap<Set<TUpdateObserver>> = new WeakMap();

// ContextManagers subscribers registry for data updates.
export const CONTEXT_SUBSCRIBERS_REGISTRY: TWorkerMap<Set<TUpdateSubscriber<any>>> = new WeakMap();

// Signal handlers registry bound to context instances.
export const CONTEXT_SIGNAL_HANDLERS_REGISTRY: TWorkerMap<TSignalListener<any>> = new WeakMap();

// Signal listeners metadata registry related to workers.
export const CONTEXT_SIGNAL_METADATA_REGISTRY: TWorkerMap<TSignalSubscriptionMetadata> = new WeakMap();

// Query listeners metadata registry related to workers.
export const CONTEXT_QUERY_METADATA_REGISTRY: TWorkerMap<TQuerySubscriptionMetadata> = new WeakMap();

// Global signal listeners registry.
export const SIGNAL_LISTENERS_REGISTRY: Set<TSignalListener<any, any>> = new Set();
