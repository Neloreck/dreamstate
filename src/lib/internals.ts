import type {
  IStringIndexed,
  TSignalListener,
  TSignalSubs,
  TSignalType,
  TUpdateObserver,
  TUpdateSubscriber
} from "./types";
import type { ContextManager } from "./management";

declare const IS_DEV: boolean;
declare const IS_DEBUG: boolean;

export const EMPTY_STRING: string = "";
export const EMPTY_ARR: Array<never> = [];
export const MANAGER_REGEX: RegExp = /Manager$/;

export const IDENTIFIER_KEY: unique symbol = Symbol(IS_DEV ? "DS_ID" : "");
export const MUTABLE_KEY: unique symbol = Symbol(IS_DEV ? "DS_MUTABLE" : "");
export const SIGNAL_LISTENER_KEY: unique symbol = Symbol(IS_DEV ? "DS_SG_SUBSCRIBER" : "");

/**
 * Store related constants.
 */

export const CONTEXT_MANAGERS_REGISTRY: IStringIndexed<ContextManager<any>> = {};
export const CONTEXT_OBSERVERS_REGISTRY: IStringIndexed<Set<TUpdateObserver>> = {};
export const CONTEXT_SUBSCRIBERS_REGISTRY: IStringIndexed<Set<TUpdateSubscriber<any>>> = {};
export const CONTEXT_STATES_REGISTRY: IStringIndexed<any> = {};

export const SIGNAL_LISTENERS: Set<TSignalListener<any>> = new Set();
export const CONTEXT_MANAGERS_SIGNAL_LISTENERS_REGISTRY: IStringIndexed<TSignalSubs> = {};

/**
 * Expose internal registry references for debugging.
 */
if (IS_DEBUG) {
  // @ts-ignore debug-only declaration.
  window.__DREAMSTATE_DEBUG_INTERNALS__ = {
    CONTEXT_MANAGERS_REGISTRY,
    CONTEXT_OBSERVERS_REGISTRY,
    CONTEXT_SUBSCRIBERS_REGISTRY,
    CONTEXT_STATES_REGISTRY,
    CONTEXT_MANAGERS_SIGNAL_LISTENERS_REGISTRY,
    SIGNAL_LISTENERS
  };
}
