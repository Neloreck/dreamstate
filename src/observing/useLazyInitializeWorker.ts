import { useEffect, useMemo } from "react";

import { EMPTY_ARR } from "../internals";
import { TAnyContextManagerConstructor, TDreamstateWorker } from "../types";
import { addWorkerObserverToRegistry, registerWorker, removeWorkerObserverFromRegistry } from "../registry";

/**
 * Initialize context manager once before tree mount and use memo.
 * Subscribe to adding/removing observers on mount/unmount.
 */
export function useLazyInitializeWorker(
  workerConstructor: TDreamstateWorker,
  updateObserver: () => void
): void {
  useMemo(function (): void {
    registerWorker(workerConstructor);
  }, EMPTY_ARR);

  useEffect(function () {
    addWorkerObserverToRegistry(workerConstructor as TAnyContextManagerConstructor, updateObserver);

    return function () {
      return removeWorkerObserverFromRegistry(workerConstructor as TAnyContextManagerConstructor, updateObserver);
    };
  }, EMPTY_ARR);
}
