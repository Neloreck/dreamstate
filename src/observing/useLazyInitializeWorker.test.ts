import { createElement, ReactElement } from "react";
import { mount } from "enzyme";

import { useLazyInitializeWorker } from "./useLazyInitializeWorker";
import { CONTEXT_OBSERVERS_REGISTRY, CONTEXT_WORKERS_REGISTRY } from "../internals";
import { unRegisterWorker } from "../testing";

import { TestContextWorker } from "@Tests/assets";

describe("useLazyInitializeWorker method functionality.", () => {
  it("Should correctly render functions and register workers.", () => {
    expect(CONTEXT_WORKERS_REGISTRY.has(TestContextWorker)).toBeFalsy();
    expect(CONTEXT_OBSERVERS_REGISTRY.has(TestContextWorker)).toBeFalsy();

    const withLazyInit = (): ReactElement => {
      useLazyInitializeWorker(TestContextWorker, jest.fn());

      return createElement("div", {}, 123);
    };

    const tree = mount(createElement(withLazyInit));

    expect(tree).toMatchSnapshot();

    expect(CONTEXT_WORKERS_REGISTRY.has(TestContextWorker)).toBeTruthy();
    expect(CONTEXT_OBSERVERS_REGISTRY.has(TestContextWorker)).toBeTruthy();
    expect(CONTEXT_OBSERVERS_REGISTRY.get(TestContextWorker)!.size).toBe(1);

    const currentWorker: TestContextWorker = CONTEXT_WORKERS_REGISTRY.get(TestContextWorker)!;
    const nextTree = mount(createElement(withLazyInit));

    expect(currentWorker).toBe(CONTEXT_WORKERS_REGISTRY.get(TestContextWorker));

    tree.unmount();
    nextTree.unmount();

    expect(CONTEXT_WORKERS_REGISTRY.get(TestContextWorker)).toBeUndefined();

    unRegisterWorker(TestContextWorker);
  });
});
