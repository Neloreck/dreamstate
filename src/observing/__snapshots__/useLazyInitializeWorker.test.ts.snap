// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useLazyInitializeWorker method functionality. Should correctly render functions and register workers. 1`] = `
<withLazyInit>
  <div>
    123
  </div>
</withLazyInit>
`;
